### DISCUSSION TITLE
How to reduce Korp memory usage and avoid running out of memory?

### DISCUSSION THREAD

**Discussion by janiemi** [spraakbanken/korp-backend#15] [Q&A]
We have recently run out of memory a couple of times on our [production Korp server](https://korp.csc.fi) at [the Language Bank of Finland](https://www.kielipankki.fi). We would be grateful for any advice or ideas on how to avoid the situation or at least make it less likely.

When the memory is low, for example `/count` returns

```json
{"type": "CQPError", "value": "[Errno 12] Cannot allocate memory"}
```

and the Korp frontend just shows the dead raven even for a very simple query, which may be rather baffling to the user.

Gunicorn and CQP would seem to consume most of the memory. We’ve been logging Korp (Gunicorn) process memory usage (maximum RSS from [`resource.getrusage`](https://docs.python.org/3/library/resource.html#resource.getrusage)), and it seems that the amount of memory used by Gunicorn workers (as reported by `resource.getrusage`) increases monotonically, that is, it would appear that allocated memory is never freed.

When I searched for “gunicorn memory leak”, I found [this discussion](https://github.com/benoitc/gunicorn/issues/1981), in which it’s suggested that a memory leak would be more likely in the actual application than in Gunicorn.

Our current (virtual) Korp server has 32 GiB RAM, and we’ve been running Gunicorn with the following parameters:

```
--worker-class gevent --workers 8 --max-requests 100 --worker-connections 250 --timeout 180 --limit-request-line 0
```

We might try to reduce the `max-requests` value even further to restart the workers more often. We might also try to investigate what kind of requests appear to increase the maximum RSS value most, or maybe something like what is presented [here](https://stackoverflow.com/a/63429528) to discover memory leaks.

The CQP child processes started by the Korp backend also seem to use rather much memory, sometimes over 1 GB for a single process, but that’s probably something you can’t really help. Nonetheless, I could perhaps try to investigate what kind of queries result in such memory usage.

In the Korp configuration, we currently have `PARALLEL_THREADS = 1`, as we’ve thought it would reduce the server load (both CPU and memory) and allow a larger number of simultaneous Korp users, at the cost of making individual queries with multiple corpora slower. But does it work like that?

Thanks in advance for any advice or hints!

**Comment by janiemi** [comment]
@mmatthiesencsc amended my message above (originally via email):

Jyrki and I lowered the number of connections and connections from:

```
--max-requests 100 --worker-connections 250
```

to:

```
--max-requests 30 --worker-connections 100
```

So gunicorn will restart after 30 requests and keep less connections per worker open.

We keep monitoring the situation.

**Comment by MartinHammarstedt** [comment]
I have a vague memory from many years ago, that leaving out the `--max-requests` parameter to Gunicorn would eventually lead to problems with Korp, which I guess would indicate some kind of memory leak. I don't remember the specifics, but I guess it wasn't investigated further due to lack of time and the fact that `--max-requests` easily solved it.

Our current (virtual) Korp server has about 50 GB RAM, and we haven't had any memory problems as far as I know. Other than having more memory, our Gunicorn parameters look a little different from yours:
```
  --worker-class gevent --workers 2 --max-requests 500 --timeout 120 --graceful-timeout 1200 --limit-request-line 0
```
The main difference is that we only have two workers. Seeing as gevent handles concurrent requests, and that most of the time-consuming work is done in subprocesses, I'm not sure if there are any big benefits from increasing the number of workers. In fact, our Korp used to run with just one worker for many years.

As for the `PARALLEL_THREADS` setting, we've had that set to `1` almost from the beginning, but I recently increased it to `2`, just as a test, and haven't noticed any problems. In theory that should speed up searching but I don't know by how much, and it could probably have a negative effect if there are too many concurrent requests.

I don't have any experience hunting memory leaks in Python, but it sounds like I might have to look into that. Unfortunately I don't have a solution to your problem, but my suggestion would be to try lowering the number of workers and see if that helps.
